Coding conventions and contributing guidelines (#272)
CROSSINGTUD/CryptoAnalysis: CODING.md

# Coding Guidelines

## General Notes

> In general we aim for keeping the coding guidelines that are common by the Java Language. This includes code style,

Reviewer: common "for"

> as well as the usage of coding patterns. Exceptions have to get discusses and agreed with the project leads.

Reviewer 2: ```suggestion
as well as the usage of coding patterns. Exceptions have to get discussed and agreed with the project leads.
```

## Checkstyle Config

> This repository contains a `checkstyl.xml` Checkstyle config which gets automatically used by the CI Runner for code 

Reviewer: checkstyle.xml

validation on the **master**, **develop** and **Pull Requests** branches.

> Feature branches do not checked because we want to things getting done quickly. This can be done if everybody codes as 

Reviewer 2: ```suggestion
Feature branches are not automatically checked because we want to get development done quickly. To achieve this, everybody may codes as 
```

they feel most familiar. A pull request will make sure `develop` and `master` branches are consistent.   

> The checkstyle config got confirmed by the core development team at UPB and IEM. Changes may only be made by the project leaders

Reviewer 2: UPB, IEM and TUD  ??

> with serious reasons.

Reviewer 2: ```suggestion
for valid reasons.
```

### Tool Support

Checkstyle is available for various IDEs.

##### Eclipse
1. Install checkstyle plugin from [here](https://checkstyle.org/eclipse-cs/)
2. In your preferences window you can select the checkstyle configuration `checkstyle.xml` from the root of the
repository.
3. Configure the Eclipse Code Formatter to automatically format the code regarding the checkstyle rules. 

> Unfortunately it does not support importing the Checkstyle Rules directly.

Reviewer 2: Capitalization of Checkstyle (Rules) is inconsistent

##### IntelliJ
1. Search for "Checkstyle-IDEA" in the `Settings->Plugin` window.
2. Install and restart the IDE.

> 3. Open settings again and select the `iem_checks.xml` file in `Settings-Checkstyle`.

Reviewer 2: Do we maybe want to rename the xml-files into something CogniCrypt-related (for clarity)?
